File: binaryTree/124_binary_tree_maximum_path_sum.go
// O(n) time | O(h) space 
func maxPathSum(root *TreeNode) int {
    ans := math.MinInt64
    calculateMaxPathSum(root, &ans)
    return int(ans)
}

func calculateMaxPathSum(node *TreeNode, ans *int) int {
    if node == nil {
        return math.MinInt64
    }

    leftMaxPathSum := MaxInt(0, calculateMaxPathSum(node.Left, ans))
    rightMaxPathSum := MaxInt(0, calculateMaxPathSum(node.Right, ans))

    if curMaxPathSum := leftMaxPathSum + rightMaxPathSum + node.Val; curMaxPathSum > *ans {
        *ans = curMaxPathSum
    }

    return MaxInt(leftMaxPathSum, rightMaxPathSum) + node.Val
}

func MaxInt(x, y int) int {
    if x > y {
        return x
    }
    return y
}

File: binaryTree/124_binary_tree_maximum_path_sum.py
class Solution:
    # O(n) time | O(h) space 
    # where n is the number of tree nodes and h is the height of tree
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        def dfs(node):
            if not node:
                return 0
            left = max(0, dfs(node.left))
            right = max(0, dfs(node.right))
            curSum = left + right + node.val
            ans[0] = max(ans[0], curSum)
            return max(left, right) + node.val

        ans = [float('-inf')]
        dfs(root)

        return ans[0] 

